include/GafferSceneUI/Private/Inspector.h

```diff
@@ -0,0 +1,238 @@
+/// optionally making node graph edits to change those properties. They allow a
+/// small set of UI components to expose editable properties without needing to
+/// know their underlying details.
+///
+/// Inspectors are responsible for _what_ is inspected, but the UI components
+/// are responsible for the Context in which inspection happens. For example, a
+/// ParameterInspector knows which parameter to inspect, but the UI provides the
+/// location to inspect via the current context.
+///
+/// Inspectors are "EditScope aware", allowing the user to create new edits using
+/// a target EditScope. One of the main contributions of the Inspector base class
+/// is to encode the rules for interaction with EditScopes.
+///
+/// Notes for future work :
+///
+/// - This is temporarily in a Private namespace so that we can iterate on it
+///   without concerns about ABI breakage. Intended to be public eventually.
+/// - We want to generalise Inspectors so they can be used with images as well
+///   as scenes. Beware any scene-centric design decisions.
+/// - We want to use a TransformInspector to replace much of TransformTool::Selection.
+///   This has additional requirements such as knowing the `transformSpace` that a node
+///   works in. We think this information can be stored in a dedicated TransformHistory
+///   class provided by SceneAlgo, avoiding any need to specialise Inspector::Result.
+class GAFFERSCENEUI_API Inspector : public IECore::RefCounted, public boost::signals::trackable
+{
+
+	public :
+
+		IE_CORE_DECLAREMEMBERPTR( Inspector );
+		IE_CORE_FORWARDDECLARE( Result );
+
+		/// The name of the property being inspected.
+		const std::string &name() const;
+
+		/// Called by the UI to inspect the property in the current context.
+		ResultPtr inspect() const;
+
+		using InspectorSignal = boost::signal<void ( Inspector * )>;
+		/// Emitted when the property queried by the inspector has changed.
+		/// The UI should use this to schedule a refresh.
+		InspectorSignal &dirtiedSignal();
+
+	protected :
+
+		Inspector( const std::string &name, const Gaffer::PlugPtr &editScope );
+
+		/// Methods to be implemented in derived classes
+		/// ============================================
+		///
+		/// The `inspect()` method delegates to several virtual methods that
+		/// should be implemented by derived classes. Inspection starts by
+		/// generating a history for the computation of the property, and then
+		/// traverses the history making additional queries at various points.
+
+		/// Must be implemented to return the history for the property being
+		/// inspected. Should return null if the property does not exist.
+		virtual GafferScene::SceneAlgo::History::ConstPtr history() const = 0;
+
+		/// Must be implemented to return the value of the property at this
+		/// point in the history. Called with `history->context` as the current
+		/// context.
+		/// \todo Perhaps this should be available directly from the history
+		/// base class?
```

Comment:
AttributeHistory uses ConstObjectPtr because that is the most general type of an attribute. But a TransformHistory could use a bare M44f. I do actually have a commit (not in this PR) that explores this a bit by refactoring much of the ParameterInspector into a `ParameterHistory SceneAlgo::parameterHistory()` method. I think it might be the way to go, but I think the point where we replace TransformTool::Selection with a TransformInspector is where it will become clear. For now I'd like to keep to the current version (which is private so we can iterate on it) for simplicity's sake.